src/adaptix/_internal/morphing/enum_provider.py

```diff
@@ -155,3 +157,85 @@ def _get_exact_value_to_member(self, enum: Type[Enum]) -> Optional[Mapping[Any,
 
     def _provide_dumper(self, mediator: Mediator, request: DumperRequest) -> Dumper:
         return _enum_exact_value_dumper
+
+
+def _extract_common_cases_from_flag(enum: Type[Flag]) -> Iterable[Flag]:
+    cases = enum.__members__.values()
+    result = []
+    for case in cases:
+        if not math.log2(case.value) % 1:
+            result.append(case)
+    return result
+
+
+class FlagProvider(BaseEnumProvider):
+    def __init__(
+        self,
+        allow_single_value: bool = False,
+        allow_duplicates: bool = True,
+        allow_compound: bool = True,
+        by_exact_value: bool = False
+    ):
+        self._allow_single_value = allow_single_value
+        self._allow_duplicates = allow_duplicates
+        self._allow_compound = allow_compound
+        self._by_exact_value = by_exact_value
+
+        self._loader = _enum_exact_value_loader if by_exact_value else _enum_name_loader
+        self._dumper = _enum_exact_value_dumper if by_exact_value else _enum_name_dumper
+
+    def _get_cases(self, enum: Type[Flag]):
+        if self._allow_compound:
+            return enum.__members__.values()
+        return _extract_common_cases_from_flag(enum)
+
+    def _flag_loader(self, data: Union[int, Iterable[Union[int, str]]], enum: Type[Flag]) -> Flag:
+        if isinstance(data, (str, int)):
+            if not self._allow_single_value:
+                raise TypeLoadError(
+                    expected_type=Iterable[str],
+                    input_value=data,
+                )
+            process_data = [data]
+        else:
+            process_data = list(data)
+
+        if not self._allow_duplicates:
+            if len(process_data) != len(set(process_data)):
+                raise ValueLoadError(
+                    msg=f"Duplicates in {enum} loader are not allowed",
+                    input_value=process_data
+                )
+
+        variants = [self._dumper(case) for case in self._get_cases(enum)]
+        bad_variants = []
+        result = enum(0)
+        for item in process_data:
+            if item not in variants:
+                bad_variants.append(item)
+                continue
+            result = result | self._loader(enum, item)
+
+        if bad_variants:
+            raise MultipleBadVariantError(
+                allowed_values=variants,
+                input_values=process_data,
+                invalid_values=bad_variants
+            )
+
+        return result
+
+    def _provide_loader(self, mediator: Mediator, request: LoaderRequest) -> Loader:
+        enum = get_type_from_request(request)
+        return partial(self._flag_loader, enum=enum)
```

Comment:
Returning partial will hold reference to provider, this prevents from producing standalone file with loader (file that knows nothing about retorts, this is planned feature).
Also permanent lookup through self attributes slow down code.

Loader should be closure capturing variables storing self atrributes